clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -869,7 +870,8 @@ private KafkaConsumer(ConsumerConfig config,
      * @param topics The list of topics to subscribe to
      * @param listener Non-null listener instance to get notifications on partition assignment/revocation for the
      *                 subscribed topics
-     * @throws IllegalArgumentException If topics is null or contains null or empty elements
+     * @throws IllegalArgumentException If topics is null or contains null or empty elements, or if listener is null
+     * @throws IllegalStateException If subscription types (topic, pattern) are mixed with each other or with assignment
```

Comment:
nit: `If subscribe is called previously with pattern, or assign is called previously.` This is to make the explanation more concrete. Ditto below.